move/axelar_gateway/sources/gateway.move

```diff
@@ -469,37 +564,474 @@ fun test_peel_proof_no_remaining_data() {
+        axelar_gateway::message::new(
+            source_chain,
+            message_id,
+            source_address,
+            destination_id,
+            payload_hash,
+        )
+    ]);
+    let data_hash = gateway_v0::approve_messages_data_hash(message_data);
+    let message_to_sign = bcs::to_bytes(&auth::new_message_to_sign(
+        domain_separator,
+        weighted_signers.hash(),
+        data_hash,
+    ));
+    let signature = axelar_gateway::proof::new_signature(
+        sui::ecdsa_k1::secp256k1_sign(keypair.private_key(), &message_to_sign, 0, true)
+    );
+    let proof = axelar_gateway::proof::create_for_testing(weighted_signers, vector[signature]);
+    let mut proof_data = bcs::to_bytes(&proof);
+    proof_data.push_back(0);
+    self.approve_messages(message_data, proof_data);
+    
+    clock.destroy_for_testing();
+    sui::test_utils::destroy(self)
+}
+
+#[test]
+fun test_rotate_signers() {
+    let ctx = &mut sui::tx_context::dummy();
+    let keypair = sui::ecdsa_k1::secp256k1_keypair_from_seed(&@0x1234.to_bytes());
+    let weighted_signers = weighted_signers::create_for_testing(
+        vector[
+            axelar_gateway::weighted_signer::new(
+                *keypair.public_key(),
+                1,
+            )
+        ],
+        1,
+        bytes32::new(@0x0),
+    );
+    let next_weighted_signers = weighted_signers::create_for_testing(
+        vector[
+            axelar_gateway::weighted_signer::new(
+                *sui::ecdsa_k1::secp256k1_keypair_from_seed(&@0x5678.to_bytes()).public_key(),
+                1,
+            )
+        ],
+        1,
+        bytes32::new(@0x1),
+    );
+
+    let operator = @0x1;
+    let domain_separator = bytes32::new(@0x2);
+    let minimum_rotation_delay = 1;
+    let previous_signers_retention = 10;
+    let initial_signers = weighted_signers;
+    let mut clock = sui::clock::create_for_testing(ctx);
+    let mut self = create_for_testing(
+        operator,
+        domain_separator,
+        minimum_rotation_delay,
+        previous_signers_retention,
+        initial_signers,
+        &clock,
+        ctx,
+    );
+
+    let message_data = bcs::to_bytes(&next_weighted_signers);
+    let data_hash = gateway_v0::rotate_signers_data_hash(message_data);
+    let message_to_sign = bcs::to_bytes(&auth::new_message_to_sign(
+        domain_separator,
+        weighted_signers.hash(),
+        data_hash,
+    ));
+    let signature = axelar_gateway::proof::new_signature(
+        sui::ecdsa_k1::secp256k1_sign(keypair.private_key(), &message_to_sign, 0, true)
+    );
+    let proof = axelar_gateway::proof::create_for_testing(weighted_signers, vector[signature]);
+
+    clock.increment_for_testing(minimum_rotation_delay);
+    self.rotate_signers(&clock, message_data, bcs::to_bytes(&proof), ctx);
+    
+    clock.destroy_for_testing();
+    sui::test_utils::destroy(self);
+}
+
+#[test]
+#[expected_failure]
```

Comment:
abort code

Reply:
again, utils does not have an abort code.